### PR TITLE
feat: delete pipeline and resources with `func delete`

### PR DESCRIPTION
**Pull request by zroubalik**
Signed-off-by: Zbynek Roubalik <zroubali@redhat.com>

# Changes

- 🎁 On cluster build: delete pipeline and all created resources with func delete

/kind enhancement

Fixes: #754

### CODE REVIEW

File: cmd/delete.go [hunk 1]
```diff
@@ -59,10 +72,11 @@ kn func delete -n apps myfunc
 `,
 		SuggestFor:        []string{"remove", "rm", "del"},
 		ValidArgsFunction: CompleteFunctionList,
-		PreRunE:           bindEnv("path", "confirm", "namespace"),
+		PreRunE:           bindEnv("path", "confirm", "namespace", "all"),
 	}
 
 	cmd.Flags().BoolP("confirm", "c", false, "Prompt to confirm all configuration options (Env: $FUNC_CONFIRM)")
+	cmd.Flags().StringP("all", "a", "true", "Delete all resources created for a function, eg. Pipelines, Secrets, etc. (Env: $FUNC_ALL) (allowed values: \"true\", \"false\")")
```
Comment (lkingland) [inline cmd/delete.go:79]: Glad to see this is set as the default :)

File: cmd/delete.go [hunk 2]
```diff
@@ -152,9 +168,37 @@ func (c deleteConfig) Prompt() (deleteConfig, error) {
...
+			Prompt: &survey.Input{
+				Message: "Function to remove:",
+				Default: deriveName(c.Name, c.Path)},
+			Validate: survey.Required,
+		},
+		{
+			Name: "all",
+			Prompt: &survey.Confirm{
+				Message: "Do you want to delete all resources?",
+				Default: true,
```
Comment (lkingland) [inline cmd/delete.go:183]:
```
Default: c.DeleteAll,
```
This _question_ default could be set to the current value of the config object's default.  That way if the user passes in the flag `--all=false`, but also `--confirm`, the value of the question will be set to the value they specified on the command line first.

File: cmd/delete.go [hunk 3]
```diff
@@ -152,9 +168,37 @@ func (c deleteConfig) Prompt() (deleteConfig, error) {
...
+	err := survey.Ask(qs, &answers)
+	if err != nil {
+		return dc, err
+	}
+
+	dc.Name = answers.Name
+	dc.DeleteAll = answers.All
+	dc.Namespace = c.Namespace
+	dc.Path = c.Path
+	dc.Verbose = c.Verbose
```
Comment (lkingland) [inline cmd/delete.go:201]: I am pretty sure this mapping to a _new_ config object is not necessary and we could just use c directoy, if the original config `c` is treated as defaults (defaults which may have been mutated by command flags before they were parsed in its constructor).

File: knative/remover.go [hunk 4]
```diff
@@ -33,8 +33,6 @@ func (remover *Remover) Remove(ctx context.Context, name string) (err error) {
 		return
 	}
 
-	fmt.Printf("Removing Knative Service: %v\n", name)
-
```
Comment (lkingland) [inline knative/remover.go:37]: ✺◟( ᐛ )◞✺

**Review by lkingland — APPROVED** [review]
Looks like a winner!

/hold so that others have a chance to take a look

File: client_test.go [hunk 5]
```diff
@@ -683,6 +683,94 @@ func TestClient_Remove_ByPath(t *testing.T) {
...
+			t.Fatalf("Expected to remove '%v', got '%v'", expectedName, name)
+		}
+		return nil
+	}
+
+	if err := client.Remove(context.Background(), fn.Function{Root: root}, deleteAll); err != nil {
+		t.Fatal(err)
+	}
+
+	if !remover.RemoveInvoked {
```
Comment (matejvasek) [inline client_test.go:720]: Isn't this condition duplicated here?

File: client_test.go [hunk 6]
```diff
@@ -683,6 +683,94 @@ func TestClient_Remove_ByPath(t *testing.T) {
...
```
Comment (matejvasek) [inline client_test.go:764]: Isn't this condition duplicated here?

File: client_test.go [hunk 7]
```diff
@@ -683,6 +683,94 @@ func TestClient_Remove_ByPath(t *testing.T) {
 
 }
 
+// TestClient_Remove_DeleteAll ensures that the remover is invoked to remove
+// and that dependent resources are removed as well -> pipeline provider is invoked
+// the Function with the name of the function at the provided root.
+func TestClient_Remove_DeleteAll(t *testing.T) {
```
Comment (matejvasek) [inline client_test.go:689]: How does the two new tests differ?

File: mock/pipelines_provider.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package mock
+
+import (
+	"context"
+
+	fn "knative.dev/kn-plugin-func"
+)
+
+type PipelinesProvider struct {
+	RunInvoked    bool
```
Comment (matejvasek) [inline mock/pipelines_provider.go:10]: I think that the callback are enough and boolean can be left out.
Users of the mock can set such bool values in theirs callbacks.

File: client_test.go (hunk 7 above)
Comment (zroubalik) [reply client_test.go:689]: Bad copy&paste.

File: client_test.go (hunk 5 above)
Comment (zroubalik) [reply client_test.go:720]: Related to the bad copy&paste

**Review by lance — COMMENTED** [review]
/lgtm

**Pull request merged**